### PR TITLE
docker-archive generates docker legacy compatible images

### PR DESCRIPTION
**Pull request by nlewo**
docker save generates image compatible with the legacy format, ie,
layers are tar, they have a configuration file and a repositories file
is created.

There are some external tools that still relie on this old format such
as mesos [1] and nixos [2].

[1] https://github.com/apache/mesos/blob//7ca46e24b3339ba27e88b99ea95362c956ef03c1/src/slave/containerizer/mesos/provisioner/docker/local_puller.cpp#L168
[2] https://github.com/NixOS/nixpkgs/blob/5c6dc717a66b6555e5757f92376527da7a553fec/pkgs/build-support/docker/default.nix#L143

### CODE REVIEW

**Comment by nlewo** [comment]
This is related to this issue https://github.com/projectatomic/skopeo/issues/425.
I only tested this patch on images coming from `docker-archive` and `docker` sources.

**Comment by nlewo** [comment]
I'm not able to get the error form the CI, and `make` locally fails : 
```
INFO[0114] Fetching latest commits from 'origin' for 'github.com/gogo/protobuf/proto' 
ERRO[0114] `git fetch -f -t origin` failed:
fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.
```
Any ideas?

**Comment by mtrmac** [comment]
The CI failure is
```
docker/tarfile/dest.go
Makefile:62: recipe for target 'validate' failed
```
Probably reformat the file using `gofmt -s -w docker/tarfile/dest.go`.  Yeah, the `Makefile` rule for `gofmt` could be improved to be clearer.

As for the local `make` failure checking out a vendored subpackage, I don’t know. Removing the `vendor` subdirectory entirely and running `make vendor` works for me.

**Comment by nlewo** [comment]
Thanks @mtrmac 
Regarding my local issue with `make`, I submitted PR #373. I also fixed several test errors in this patch but I still have an test error with Skopeo.
The error is related to `dir` and `docker` source/dest which should not be impacted by this patch:
```
Running skopeo --debug copy docker://busybox dir:/tmp/copy-1238023601
Running skopeo --tls-verify=false --debug copy dir:/tmp/copy-1238023601 docker://localhost:5555/busybox:unsigned
Running skopeo --tls-verify=false --debug copy docker://localhost:5555/busybox:unsigned dir:/tmp/copy-2608739164
Running diff -urN /tmp/copy-1238023601 /tmp/copy-2608739164
# the diff should be nil
```
Moreover, I cannot reproduce it when I use the docker registry.

Is there a trick with the test registry? (maybe it is using `tarfile`)
How could I get the debug messages of Skopeo in the test output?

**Comment by nlewo** [comment]
Ok, I've finally fixed tests

File: copy/copy.go [hunk 1]
```diff
@@ -609,6 +609,18 @@ func (ic *imageCopier) copyBlobFromStream(srcStream io.Reader, srcInfo types.Blo
 		inputInfo.Size = -1
 	}
 
+	// === Decompress the layer if it is compressed and decompression is desired
+	// This is currently only used by docker-archive
+	if isCompressed && ic.dest.ShouldDecompressLayers() {
+		logrus.Debugf("Blob will be decompressed")
+		destStream, err = decompressor(destStream)
+		if err != nil {
```
Comment (mtrmac) [inline copy/copy.go]: There is already a decompression code around `getOriginalLayerCopyWriter`. If at all possible, we should only decompress the data once and share the output between the two consumers, if they exist at the same time.

File: docker/tarfile/dest.go [hunk 2]
```diff
@@ -122,18 +130,49 @@ func (d *Destination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types
...
+			digester := digest.Canonical.Digester()
+			_, err = io.Copy(digester.Hash(), streamCopy)
+			if err != nil {
+				return types.BlobInfo{}, err
+			}
+			_, err = streamCopy.Seek(0, os.SEEK_SET)
+			if err != nil {
+				return types.BlobInfo{}, err
+			}
+			inputInfo.Digest = digester.Digest()
```
Comment (mtrmac) [inline docker/tarfile/dest.go]: It would probably be cheaper to either compute the digest always, or use `io.TeeReader`/`io.MultiWriter`, so that we make as few passes over the data as possible.

File: types/types.go [hunk 3]
```diff
@@ -94,6 +94,7 @@ type ImageReference interface {
 type BlobInfo struct {
 	Digest      digest.Digest // "" if unknown.
 	Size        int64         // -1 if unknown
+	Config      bool          // true if the blob is a config blob
```
Comment (mtrmac) [inline types/types.go]: This is not very elegant... conceptually, the information is already available in `MediaType`, and more importantly if `Config:true` is only set by `Image.ConfigInfo()`, the caller structurally knows that the blob is a config without having to ask.

Maybe we should have separate `ImageDestination.{PutLayer,PutBlob}`, or at the very least least move this `Config` bool into a separate parameter of `PutBlob` which can be set by the caller directly instead of relying on `ConfigInfo`.  (@runcom WDYT?)

File: types/types.go [hunk 4]
```diff
@@ -149,6 +150,8 @@ type ImageDestination interface {
 	SupportsSignatures() error
 	// ShouldCompressLayers returns true iff it is desirable to compress layer blobs written to this destination.
 	ShouldCompressLayers() bool
+	// ShouldDecompressLayers returns true iff it is desirable to decompress layer blobs written to this destination.
+	ShouldDecompressLayers() bool
```
Comment (mtrmac) [inline types/types.go]: Please use a single `enum` (with cases compress/decompress/preserveoriginal) instead of two booleans, to make structurally sure `ShouldCompressLayers`&&`ShouldDecompressLayers` can’t be true at the same time.

File: types/types.go (hunk 3 above)
Comment (mtrmac) [reply types/types.go]: …separate `imageDestination.{PutLayer,PutConfig}` that is.

Comment (nlewo) [reply types/types.go]: If this is currently the only need to separate layers from the config, I think it would be less intrusive to add a bool paramater to the putBlob method.
Do you think there are other use cases that need this distinction?

File: types/types.go (hunk 4 above)
Comment (nlewo) [reply types/types.go]: Ok.

File: copy/copy.go (hunk 1 above)
Comment (nlewo) [reply copy/copy.go]: @mtrmac `getOriginalLayerCopyWriter` is only used when the format of the source image is Docker SchemaV1. So, we could have two decompression if the source format image is Docker SchemaV1 and the destination is `docker-archive`.
But I think it's not trivial to change this behavior since getOriginalLayerCopyWriter is already a little bit tricky...
Do you know if SchemaV1 is still widely used?

**Comment by nlewo** [comment]
CI is failing because Skopeo needs to be patched. Someone could still have a look to this PR?

**Comment by gilligan** [comment]
I'm a nixos / docker tooling user and I would very much love to see this merged :)

**Comment by gilligan** [comment]
@nlewo @mtrmac what is missing to move this forward? what can be done to support this?

**Comment by nlewo** [comment]
This PR has never been really reviewed. I have had some comments from @mtrmac but I don't know if this "feature" could be merged.
One month ago, I've rebased it on master and this was not trivial:/ So, now I don't know what to do with this.
But of course, I would be happy to reply to reviews!

**Comment by mtrmac** [comment]
Sorry about that, I was unavailable for most of December and getting to this through my backlog took a while.

Here’s an in-progress review, I’m afraid I have to context switch to something else for the rest of today…

File: copy/copy.go [hunk 5]
```diff
@@ -666,6 +667,18 @@ func (c *copier) copyBlobFromStream(srcStream io.Reader, srcInfo types.BlobInfo,
 		inputInfo.Size = -1
 	}
 
+	// === Decompress the layer if it is compressed and decompression is desired
+	// This is currently only used by docker-archive
+	if isCompressed && c.dest.CompressesLayers() == types.Decompress {
```
[the remark this answers is not in the document]
Comment (nlewo) [reply copy/copy.go]: Exactly, the sequence can be rewritten such as:
```
if compression is desired and it is possible
    compress()
else if decompression is desired and it is possible
    decompress()
else
    continue
```

File: copy/copy.go [hunk 6]
```diff
@@ -606,7 +606,7 @@ func computeDiffID(stream io.Reader, decompressor compression.DecompressorFunc)
 // and returns a complete blobInfo of the copied blob.
 func (c *copier) copyBlobFromStream(srcStream io.Reader, srcInfo types.BlobInfo,
 	getOriginalLayerCopyWriter func(decompressor compression.DecompressorFunc) io.Writer,
-	canCompress bool) (types.BlobInfo, error) {
+	canCompress bool, isConfig bool) (types.BlobInfo, error) {
```
[the remark this answers is not in the document]
Comment (nlewo) [reply copy/copy.go]: Ok.

File: docker/tarfile/types.go [hunk 7]
```diff
@@ -12,7 +12,7 @@ const (
 	manifestFileName = "manifest.json"
 	// legacyLayerFileName        = "layer.tar"
 	// legacyConfigFileName       = "json"
-	// legacyVersionFileName      = "VERSION"
+	legacyVersionFileName = "VERSION"
 	// legacyRepositoriesFileName = "repositories"
```
[the remark this answers is not in the document]
Comment (nlewo) [reply docker/tarfile/types.go]: I use these constants.

File: docker/tarfile/dest.go [hunk 8]
```diff
@@ -91,20 +94,20 @@ func (d *Destination) MustMatchRuntimeOS() bool {
...
+		}
+		if ok {
+			return types.BlobInfo{Digest: inputInfo.Digest, Size: size}, nil
+		}
 	}
 
-	if inputInfo.Size == -1 { // Ouch, we need to stream the blob into a temporary file just to determine the size.
+	// Ouch, we need to stream the blob into a temporary file just to determine the size.
+	// When the layer is decompressed, we also have to generate the digest on uncompressed datas.
+	if inputInfo.Size == -1 {
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: You’re right that the `copy.Image` code won’t have `inputInfo.Digest == "" && inputInfo.Size != -1`, but the documented behavior of the interface does allow it.

The added check below (`Can not stream a blob with unknown digest to docker tarfile`) is too late; check in this condition, and then at the location of the current a digest ill always be available.

File: docker/tarfile/dest.go [hunk 9]
```diff
@@ -113,10 +116,15 @@ func (d *Destination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types
...
-		size, err := io.Copy(streamCopy, stream)
+		digester := digest.Canonical.Digester()
+		tee := io.TeeReader(stream, digester.Hash())
+		size, err := io.Copy(streamCopy, tee)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+		if inputInfo.Digest == "" {
+			inputInfo.Digest = digester.Digest()
+		}
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: Yes, this. Thanks.

File: copy/copy.go (hunk 1 above)
Comment (mtrmac) [reply copy/copy.go]: schema1 is still around (and old images will keep using it), but hopefully its will diminish over time… just like using the `repositories` archive format :)

I really don’t like decompressing the possibly gigabytes of data twice, but you’re right that the code is tricky and I don’t quite have the extra time necessary to get this right, so I guess let’s live with the two separate decompressions.

File: docker/tarfile/dest.go [hunk 10]
```diff
@@ -176,13 +207,52 @@ func (d *Destination) PutManifest(m []byte) error {
...
+
+		// The legacy format requires a config file per layer
+		layerConfig := make(map[string]interface{})
+		id := l.Digest.Hex()
+		layerConfig["id"] = id
+
+		// All layers expect the root one have a parent
+		if i != 0 {
+			layerConfig["parent"] = man.LayersDescriptors[i-1].Digest.Hex()
```
[the remark this answers is not in the document]
Comment (nlewo) [reply docker/tarfile/dest.go]: I think this could be implemented but I don't have any image to test this logic. Do you know how could I produce this kind of image or where I could find one?

**Comment by nlewo** [comment]
Thanks for your review. I've addressed all of your comments, excepted the one related to the `ChainID` which is still in discussion.

File: docker/tarfile/dest.go [hunk 11]
```diff
@@ -193,7 +226,47 @@ func (d *Destination) PutManifest(m []byte) error {
 	}
 
 	// FIXME? Do we also need to support the legacy format?
-	return d.sendFile(manifestFileName, int64(len(itemsBytes)), bytes.NewReader(itemsBytes))
+	return d.sendBytes(manifestFileName, itemsBytes)
+}
+
+// WriteLegacyLayerMetadata writes legacy VERSION and configuration files for all layers
+func (d *Destination) WriteLegacyLayerMetadata(layerDescriptors []manifest.Schema2Descriptor) (layerPaths []string, err error) {
```
Comment (mtrmac) [inline docker/tarfile/dest.go]: Nit: please make this method private

File: docker/tarfile/dest.go [hunk 12]
```diff
@@ -122,17 +115,42 @@ func (d *Destination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types
...
+		}
 		stream = streamCopy
 		logrus.Debugf("... streaming done")
 	}
 
-	digester := digest.Canonical.Digester()
-	tee := io.TeeReader(stream, digester.Hash())
-	if err := d.sendFile(inputInfo.Digest.String(), inputInfo.Size, tee); err != nil {
-		return types.BlobInfo{}, err
+	if inputInfo.Digest.String() == "" {
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: Right, at this point the code can unconditionally assume `inputInfo.Digest` is valid.

The start of `PutBlob` could, I think, be
```
func (d *Destination) PutBlob(stream io.Reader, inputInfo types.BlobInfo, isConfig bool) (types.BlobInfo, error) {
	// Ouch, we need to stream the blob into a temporary file just to determine the size or digest.
	if inputInfo.Size == -1 || inputInfo.Digest.String() == "" {
 		logrus.Debugf("docker tarfile: input with unknown size or digest, streaming to disk first ...")
```
and transparently handle the hypothetical case of known size && unknown digest instead of failing.

File: docker/tarfile/dest.go [hunk 13]
```diff
@@ -122,17 +115,42 @@ func (d *Destination) PutBlob(stream io.Reader, inputInfo types.BlobInfo) (types
...
+		}
+		d.config = buf
+		if err := d.sendFile(inputInfo.Digest.Hex()+".json", inputInfo.Size, bytes.NewReader(buf)); err != nil {
+			return types.BlobInfo{}, errors.Wrap(err, "Error writing Config file")
+		}
+	} else {
+		if err := d.sendFile(filepath.Join(inputInfo.Digest.Hex(), "layer.tar"), inputInfo.Size, stream); err != nil {
+			return types.BlobInfo{}, err
+		}
+		d.blobs[inputInfo.Digest] = types.BlobInfo{Digest: inputInfo.Digest, Size: inputInfo.Size}
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: So, why not move the `d.blobs[inputInfo.Digest = …` line one line lower, out of the `if`?

File: docker/tarfile/dest.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: I guess, take any existing image (e.g. `skopeo copy docker://whatever dir:tmpdir` to have an easy way to edit it), edits its manifest to repeat one item of the `layers` array, edit its `config.json` to correspondingly list the layer twice in `rootfs.diff_ids`, and update the manifest with the updated manifest digest.

(To be honest, this would probably be the first time anyone is actually testing this corner case, so something else might turn out to be broken.)

**Comment by mtrmac** [comment]
The `FAIL - does not have a valid DCO` test failures: Please sign your commits (`git commit --amend -s`)

**Comment by gilligan** [comment]
@nlewo yay, keep those fixups coming! :heart:

File: docker/tarfile/dest.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nlewo) [reply docker/tarfile/dest.go]: Thx. I think I successfully built an image such like that. Currently, when I docker-load this image (after copying it with `skopeo docker-archive`), Docker sees 5 layers while there are 6 layers in the `rootfs` array.
I will see how we could handle this kind of images.

[the remark this answers is not in the document]
Comment (nlewo) [reply docker/tarfile/dest.go]: @mtrmac I've modified an image in order to reuse a layer by using the skopeo transport `dir`. I've tryied to copy this image to `docker-daemon` and it prints:
```
Skipping fetch of repeat blob sha256:6fbc4a21b806838b63b774b338c6ad19d696a9e655f50b4e358cc4006c3baa79
```
In fact, the same behavior is observed if we use the proposed `docker-archive` implementation: the duplicated layer is ignored.

To be honest, I don't really know if the generated image reproduces well this use case and I'm not sure what is the expected behavior when this kind of images are loaded by Docker... It would be easier by having a real image.

Moreover, in the Docker implementation, they use the `chainID` to generate the path of the layer in the tar stream. This is hard for us to do this because the layer is sent to the tar stream by the `PutBlob` method which is designed to handle each layer independently from other ones: we can not easily use previous layer sha to generate the current layer path.

Finally, it seems this kind of image would probably never been used!
So instead of adding complexity to the code for this corner case, we could print a message that says "Repeated layer is ignored".

[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: > Skipping fetch of repeat blob sha256:…

That part is an optional optimization to `PutBlob` and more or less known to work; I am more worried about the correctness of the generated metadata (in this case, primarily the `parent` links).

(I need to test this scenario out myself, maybe I am talking nonsense; hopefully tomorrow. For now just this quick note.)

**Comment by nlewo** [comment]
@gilligan @kuznero I pushed a [nixpkgs branch](https://github.com/nlewo/nixpkgs/tree/skopeo-legacy-image) where nixpkgs Docker tools uses this version of Skopeo. It works well on nixpkgs provided docker examples, but it would be nice if you can also try it on your images:)
You can also easily use it : 
```
nix-build -I nixpkgs=https://github.com/nlewo/nixpkgs/archive/0b67a3bbeaf16ca274fb467e45021ce98aad4266.tar.gz '<nixpkgs>' -A dockerTools.examples.onTopOfPulledImage

**Comment by gilligan** [comment]
@nlewo awesome! Thank you. I will try to test it as soon as possible and will report back here!

**Comment by gilligan** [comment]
@nlewo i had a quick look, we don’t really use pullImage as we build everything from scratch but I definitely want to try your skopeo branch/rev that you are setting in the latest commit to test if pushing the archive built with dockerTools.buildImage to the registry works. It should, right?

**Comment by nlewo** [comment]
@gilligan yes, it should work with and without this patch (i'm already using skopeo to push images that have been produced with `dockerTools.buildImage`).

**Comment by gilligan** [comment]
huh? Last I tried that did not work since skopeo was complaining about a missing manifest file I believe

**Comment by nlewo** [comment]
@gilligan I'm pushing to a registry several images generated by `dockerTools.buildImage` without any troubles. Can you share your image or provide more details?
I think it's not the right place here to discuss about this. Can you open an issue or ping me on the irc `nixos` channel (my nick is `lewo`).

File: docker/tarfile/dest.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: OK, tested this. TL;DR: the _legacy metadata_ in the generated tar file crashes the Docker daemon on `docker load`, so this really needs to be fixed. 

(To be fair, the unmodified tar file which includes the new metadata as well works fine with `docker load`; but generating of the legacy format is the whole port of this PR, so the legacy format should be correct.)

```console
# Build skopeo with this PR, updating layers.go as necessary
$ ./skopeo --override-os linux --policy default-policy.json copy docker://busybox dir:edits
$ python -mjson.tool < edits/5b0d59026729b68570d99bc4f3f7c31a2e4f2a5736435641565d93e7c25bd2c3.tar > _
# To duplicate the `rootfs.diff_ids` array entry
$ vi _
$ shasum --algorithm 256 _ | cut -d ' ' -f 1
e2f0703c8b534d097a822e895b0852ceedccb83a7a122aa25f92757b0ee1fcc3
$ ls -l _
$ mv _ edits/$(shasum --algorithm 256 _ | cut -d ' ' -f 1).tar
# To update the size and digest of the config, and duplicate the layer entry
$ vi edits/manifest.json
# Test that the result is minimally consistent
$ ./skopeo --override-os linux --policy default-policy.json copy dir:edits dir:t
$ rm -rf t
$ ./skopeo --override-os linux --policy default-policy.json copy dir:edits docker-archive:archive0.tar:busybox:v0
$ mkdir archive0
$ cd archive0/
$ tar xf ../archive0.tar 
$ ls
4febd3792a1fb2153108b4fa50161c6ee5e3d16aa483a63215f936a113a88e9a	manifest.json
e2f0703c8b534d097a822e895b0852ceedccb83a7a122aa25f92757b0ee1fcc3.json	repositories
# The layer refers to itself as a parent!
$ python -mjson.tool 4febd3792a1fb2153108b4fa50161c6ee5e3d16aa483a63215f936a113a88e9a/json
…
    "id": "4febd3792a1fb2153108b4fa50161c6ee5e3d16aa483a63215f936a113a88e9a",
…
    "parent": "4febd3792a1fb2153108b4fa50161c6ee5e3d16aa483a63215f936a113a88e9a"
…
# Make sure the legacy loader is used
$ rm manifest.json
override r--r--r--  mitr/staff for manifest.json? y
$ tar cf ../archive1.tar .
$ cd ..
# Verify that skopeo can’t use a legacy-only archive
$ ./skopeo --policy default-policy.json copy docker-archive:archive1.tar:busybox:v0 dir:t
WARN[0000] docker-archive: references are not supported for sources (ignoring) 
FATA[0000] Error determining manifest MIME type for docker-archive:archive1.tar:docker.io/library/busybox:v0: Error loading tar component manifest.json: file does not exist 
# Try using the file with docker. This hangs for a while, eventually failing with
# docker load < ~mitr/archive1.tar
error during connect: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.26/images/load?quiet=0: EOF
```

and the system logs show the daemon aborting with a stack overflow due to an infinite recursion in `github.com/docker/docker/image/tarexport.(*tarexporter).legacyLoadImage()`.

[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: > Moreover, in the Docker implementation, they use the `chainID` to generate the path of the layer in the tar stream. This is hard for us to do this because the layer is sent to the tar stream by the `PutBlob` method which is designed to handle each layer independently from other ones: we can not easily use previous layer sha to generate the current layer path.

[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/tarfile/dest.go]: > Moreover, in the Docker implementation, they use the `chainID` to generate the path of the layer in the tar stream. This is hard for us to do this because the layer is sent to the tar stream by the `PutBlob` method which is designed to handle each layer independently from other ones: we can not easily use previous layer sha to generate the current layer path.

The metadata is created not in `PutBlob`, but in `writeLegacyLayerMetadata`, at which point the full ordered set of layers is available as `layerDescriptors`.

**Comment by nlewo** [comment]
@mtrmac Thanks for your how to "build crazy images" :) When I had triyed, I didn't remove the manifest.json file. This explains why I was working for me.

I pushed a commit that avoids the loop by skipping repeated  layers. A warning is printed to inform the user.

**Comment by mtrmac** [comment]
> I pushed a commit that avoids the loop by skipping repeated layers. 

That changes the content of the image; the duplicates matter (see below for an example). Please either compute a `ChainID` as discussed above, of generate completely new IDs which are unique within the tarball and unrelated to the layer `DiffID` values.

---

Actually, there is a far simpler way to produce such an image: Create an empty `empty` file, and `Dockerfile` with
```
FROM fedora

ADD empty /
RUN rm /empty
ADD empty /
```
This results in the following `RootFS.Layers`:
```json
                "sha256:d32459d9ce237564fb93573b85cbc707600d43fbe5e46e8eeef22cad914bb516",
                "sha256:4b5c23136932e4a87cecc06f674ff3f66ca21c8a61653b97aea095b5822ded60",
                "sha256:46c039e328386fe951548c428c030f15bf6c764aa3222fedb9d114d2b3270bb8",
                "sha256:4b5c23136932e4a87cecc06f674ff3f66ca21c8a61653b97aea095b5822ded60"
```

i.e. the two `ADD empty /` directives cause two exactly identical layers to be created — and merely removing the last one changes the image, from an image which contains `/empty` to one which doesn’t.

**Comment by mtrmac** [comment]
Would https://github.com/mtrmac/image/tree/docker-archive-chainid (testable via https://github.com/mtrmac/skopeo/tree/docker-archive-chainid ) work for you? At least `docker load` accepts both the new and old metadata from the generated tarball, with a
```
FROM fedora

ADD a /
ADD a /
ADD b /
ADD c /
ADD b /
ADD c /
ADD a /
```

(Note how the layers end up in the root of the tarball again; the legacy paths all become symlinks.)

**Comment by nlewo** [comment]
@mtrmac thanks for your patches and the skopeo branch:) 
Your patches work well on my docker images.

**Comment by nlewo** [comment]
@mtrmac I've integrated your patches and rebased this PR on master. 
Are there still some blocking points?

**Comment by mtrmac** [comment]
👍 Thanks!

@runcom PTAL

**Comment by mtrmac** [comment]
Failing `make test-skopeo` in this PR should be fixed through https://github.com/projectatomic/skopeo/pull/480 .

**Comment by nlewo** [comment]
@runcom Could you please have a look?

**Comment by nlewo** [comment]
@mtrmac it would be really nice to get this PR merged since I'm waiting for it to improve our Docker tooling. How could we move forward?

**Comment by mtrmac** [comment]
@runcom ping again, this has been waiting for you for 24 days now.

**Comment by runcom** [comment]
@nlewo this needs one last rebase - I'm reviewing it

**Comment by runcom** [comment]
LGTM

**Comment by nlewo** [comment]
I've rebased and made some tests on my side.
Thanks @mtrmac @runcom
